File: lib/topology.py
# Copyright 2014 ETH Zurich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
:mod:`topology` --- SCION topology parser
===========================================
"""
# Stdlib
import logging
from sys import maxsize

# SCION
from lib.defines import (
    BEACON_SERVICE,
    CERTIFICATE_SERVICE,
    PATH_SERVICE,
    ROUTER_SERVICE,
    SIBRA_SERVICE,
)
from lib.errors import SCIONKeyError
from lib.packet.host_addr import haddr_parse_interface
from lib.packet.scion_addr import ISD_AS
from lib.types import LinkType
from lib.util import load_json_file


class Element(object):
    """
    The Element class is the base class for elements specified in the topology
    file.

    :ivar HostAddrBase addr: Host address of a server or border router.
    :ivar str name: element name or id
    """
    def __init__(self, public=None, bind=None, name=None):
        """
        :param str addr: (addr_type, address) of the element's Host address.
        :param str name: element name or id
        """
        self.public = self._parse_addrs(public)
        self.bind = self._parse_addrs(bind)
        self.name = None
        if name is not None:
            self.name = str(name)

    def _parse_addrs(self, value):
        if not value:
            return []
        addrs = []
        if not isinstance(value, (list, tuple)):
            value = [value]
        for val in value:
            addrs.append((haddr_parse_interface(val['Addr']), val['L4Port']))
        return addrs


class ServerElement(Element):
    """The ServerElement class represents one of the servers in the AS."""
    def __init__(self, server_dict, name=None):  # pragma: no cover
        """
        :param dict server_dict: contains information about a particular server.
        :param str name: server element name or id
        """
        super().__init__(server_dict['Public'], server_dict.get('Bind'), name)


class InterfaceElement(Element):
    """
    The InterfaceElement class represents one of the interfaces of an border
    router.

    :ivar int if_id: the interface ID.
    :ivar int isd_as: the ISD-AS identifier of the neighbor AS.
    :ivar str link_type: the type of relationship to the neighbor AS.
    :ivar int to_udp_port:
        the port number receiving UDP traffic on the other end of the link.
    :ivar int udp_port: the port number used to send UDP traffic.
    """
    def __init__(self, if_id, interface_dict, name=None):
        """
        :param dict interface_dict: contains information about the interface.
        """
        self.internaladdridx = int(interface_dict['InternalAddrIdx'])
        self.overlay = interface_dict['Overlay']
        self.bandwidth = int(interface_dict['Bandwidth'])
        self.isd_as = ISD_AS(interface_dict['ISD_AS'])
        self.link_type = interface_dict['LinkType']
        self.mtu = int(interface_dict['MTU'])
        self.remote = self._parse_addrs(interface_dict['Remote'])
        self.if_id = if_id
        self.to_if_id = 0  # Filled in later by IFID packets
        super().__init__(interface_dict['Public'], interface_dict.get('Bind'),
                         name)


class RouterElement(object):
    """
    The RouterElement class represents one of the border routers.
    """
    def __init__(self, router_dict, name=None):  # pragma: no cover
        """
        :param dict router_dict: contains information about an border router.
        :param str name: router element name or id
        """
        self.name = name
        self.internaladdrs = []
        for attr in router_dict['InternalAddrs']:
            self.internaladdrs.append(Element(attr['Public'], attr.get('Bind')))
        self.interfaces = {}
        self._order_id = maxsize
        for if_id, intf in router_dict['Interfaces'].items():
            if_id = int(if_id)
            self.interfaces[if_id] = InterfaceElement(if_id, intf)
            if if_id < self._order_id:
                self._order_id = if_id

    def __lt__(self, other):  # pragma: no cover
        return self._order_id < other._order_id


class Topology(object):
    """
    The Topology class parses the topology file of an AS and stores such
    information for further use.

    :ivar bool is_core_as: tells whether an AS is a core AS or not.
    :ivar ISD_AS isd_is: the ISD-AS identifier.
    :ivar list beacon_servers: beacons servers in the AS.
    :ivar list certificate_servers: certificate servers in the AS.
    :ivar list path_servers: path servers in the AS.
    :ivar list parent_border_routers:
        border routers linking the AS to its parents.
    :ivar list child_border_routers:
        border routers linking the AS to its children.
    :ivar list peer_border_routers: border router linking the AS to its peers.
    :ivar list core_border_routers:
        border router linking the core AS to another core AS.
    """
    def __init__(self):  # pragma: no cover
        self.is_core_as = False
        self.isd_as = None
        self.mtu = None
        self.beacon_servers = []
        self.certificate_servers = []
        self.path_servers = []
        self.sibra_servers = []
        self.parent_border_routers = []
        self.child_border_routers = []
        self.peer_border_routers = []
        self.core_border_routers = []
        self.zookeepers = []

    @classmethod
    def from_file(cls, topology_file):  # pragma: no cover
        """
        Create a Topology instance from the file.

        :param str topology_file: path to the topology file
        """
        return cls.from_dict(load_json_file(topology_file))

    @classmethod
    def from_dict(cls, topology_dict):  # pragma: no cover
        """
        Create a Topology instance from the dictionary.

        :param dict topology_dict: dictionary representation of a topology
        :returns: the newly created Topology instance
        :rtype: :class:`Topology`
        """
        topology = cls()
        topology.parse_dict(topology_dict)
        return topology

    def parse_dict(self, topology):
        """
        Parse a topology dictionary and populate the instance's attributes.

        :param dict topology: dictionary representation of a topology
        """
        self.is_core_as = topology['Core']
        self.isd_as = ISD_AS(topology['ISD_AS'])
        self.mtu = topology['MTU']
        self._parse_srv_dicts(topology)
        self._parse_router_dicts(topology)
        self._parse_zk_dicts(topology)

    def _parse_srv_dicts(self, topology):
        for type_, list_ in (
            ("BeaconService", self.beacon_servers),
            ("CertificateService", self.certificate_servers),
            ("PathService", self.path_servers),
            ("SibraService", self.sibra_servers),
        ):
            for k, v in topology[type_].items():
                list_.append(ServerElement(v, k))

    def _parse_router_dicts(self, topology):
        for k, v in topology['BorderRouters'].items():
            router = RouterElement(v, k)
            ntype_map = {
                LinkType.PARENT: self.parent_border_routers,
                LinkType.CHILD: self.child_border_routers,
                LinkType.PEER: self.peer_border_routers,
                LinkType.CORE: self.core_border_routers,
            }
            for intf in router.interfaces.values():
                ntype_map[intf.link_type].append(router)

    def _parse_zk_dicts(self, topology):
        for zk in topology['ZookeeperService'].values():
            haddr = haddr_parse_interface(zk['Addr'])
            zk_host = "[%s]:%s" % (haddr, zk['L4Port'])
            self.zookeepers.append(zk_host)

    def get_all_border_routers(self):
        """
        Return all border routers associated to the AS.

        :returns: all border routers associated to the AS.
        :rtype: list
        """
        all_border_routers = []
        all_border_routers.extend(self.parent_border_routers)
        all_border_routers.extend(self.child_border_routers)
        all_border_routers.extend(self.peer_border_routers)
        all_border_routers.extend(self.core_border_routers)
        return all_border_routers

    def get_own_config(self, server_type, server_id):
        type_map = {
            BEACON_SERVICE: self.beacon_servers,
            CERTIFICATE_SERVICE: self.certificate_servers,
            PATH_SERVICE: self.path_servers,
            ROUTER_SERVICE: self.get_all_border_routers(),
            SIBRA_SERVICE: self.sibra_servers,
        }
        try:
            target = type_map[server_type]
        except KeyError:
            logging.critical("Unknown server type: \"%s\"", server_type)
            raise SCIONKeyError from None

        for i in target:
            if i.name == server_id:
                return i
        else:
            logging.critical("Could not find server: %s", server_id)
            raise SCIONKeyError from None

File: go/lib/topology/topology.go
// Copyright 2016 ETH Zurich
// Copyright 2018 ETH Zurich, Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package topology

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net"
	"sort"
	"time"

	"github.com/scionproto/scion/go/lib/addr"
	"github.com/scionproto/scion/go/lib/common"
	"github.com/scionproto/scion/go/lib/serrors"
	jsontopo "github.com/scionproto/scion/go/lib/topology/json"
	"github.com/scionproto/scion/go/lib/topology/underlay"
)

// EndhostPort is the underlay port that the dispatcher binds to on non-routers.
const EndhostPort = underlay.EndhostPort

// ErrAddressNotFound indicates the address was not found.
var ErrAddressNotFound = serrors.New("address not found")

type (
	// RWTopology is the topology type for applications and libraries that need write
	// access to AS topology information (e.g., discovery, topology reloaders).
	//
	// The first section contains metadata about the topology. All of these fields
	// should be self-explanatory. The unit of TTL is seconds, with the zero value
	// indicating an infinite TTL.
	//
	// The second section concerns the Border routers. BRNames is just a sorted slice
	// of the names of the BRs in this topology. Its contents is exactly the same as
	// the keys in the BR map.
	//
	// The BR map points from border router names to BRInfo structs, which in turn
	// are lists of IFID type slices, thus defines the IFIDs that belong to a
	// particular border router. The IFInfoMap points from interface IDs to IFInfo structs.
	//
	// The third section in RWTopology concerns the SCION-specific services in the topology.
	// The structure is identical between the various elements. For each service,
	// there is again a sorted slice of names of the servers that provide the service.
	// Additionally, there is a map from those names to TopoAddr structs.
	RWTopology struct {
		Timestamp  time.Time
		IA         addr.IA
		Attributes []jsontopo.Attribute
		MTU        int

		BR        map[string]BRInfo
		BRNames   []string
		IFInfoMap IfInfoMap

		CS                        IDAddrMap
		DS                        IDAddrMap
		HiddenSegmentLookup       IDAddrMap
		HiddenSegmentRegistration IDAddrMap
		SIG                       map[string]GatewayInfo
	}

	// GatewayInfo describes a scion gateway.
	GatewayInfo struct {
		CtrlAddr        *TopoAddr
		DataAddr        *net.UDPAddr
		AllowInterfaces []uint64
	}

	// BRInfo is a list of AS-wide unique interface IDs for a router. These IDs are also used
	// to point to the specific internal address clients should send their traffic
	// to in order to use that interface, via the IFInfoMap member of the Topo
	// struct.
	BRInfo struct {
		Name string
		// CtrlAddrs are the local control-plane addresses.
		CtrlAddrs *TopoAddr
		// InternalAddr is the local data-plane address.
		InternalAddr *net.UDPAddr
		// IFIDs is a sorted list of the interface IDs.
		IFIDs []common.IFIDType
		// IFs is a map of interface IDs.
		IFs map[common.IFIDType]*IFInfo
	}

	// IfInfoMap maps interface ids to the interface information.
	IfInfoMap map[common.IFIDType]IFInfo

	// IFInfo describes a border router link to another AS, including the internal data-plane
	// address applications should send traffic to and information about the link itself and the
	// remote side of it.
	IFInfo struct {
		// ID is the interface ID. It is unique per AS.
		ID           common.IFIDType
		BRName       string
		CtrlAddrs    *TopoAddr
		Underlay     underlay.Type
		InternalAddr *net.UDPAddr
		Local        *net.UDPAddr
		Remote       *net.UDPAddr
		RemoteIFID   common.IFIDType
		Bandwidth    int
		IA           addr.IA
		LinkType     LinkType
		MTU          int
		BFD          BFD
	}

	// IDAddrMap maps process IDs to their topology addresses.
	IDAddrMap map[string]TopoAddr

	// TopoAddr wraps the possible addresses of a SCION service and describes
	// the underlay to be used for contacting said service.
	TopoAddr struct {
		SCIONAddress    *net.UDPAddr
		UnderlayAddress *net.UDPAddr
	}

	// BFD is the configuration for a BFD session
	BFD struct {
		Disable               bool
		DetectMult            uint8
		DesiredMinTxInterval  time.Duration
		RequiredMinRxInterval time.Duration
	}
)

// NewRWTopology creates new empty Topo object, including all possible service maps etc.
func NewRWTopology() *RWTopology {
	return &RWTopology{
		BR:                        make(map[string]BRInfo),
		CS:                        make(IDAddrMap),
		DS:                        make(IDAddrMap),
		HiddenSegmentLookup:       make(IDAddrMap),
		HiddenSegmentRegistration: make(IDAddrMap),
		SIG:                       make(map[string]GatewayInfo),
		IFInfoMap:                 make(IfInfoMap),
	}
}

// RWTopologyFromJSONTopology converts a parsed JSON struct topology to a topology usable by Go
// code.
func RWTopologyFromJSONTopology(raw *jsontopo.Topology) (*RWTopology, error) {
	t := NewRWTopology()
	if err := t.populateMeta(raw); err != nil {
		return nil, err
	}
	if err := t.populateBR(raw); err != nil {
		return nil, err
	}
	if err := t.populateServices(raw); err != nil {
		return nil, err
	}
	return t, nil
}

// RWTopologyFromJSONBytes extracts the topology from a JSON representation in raw byte format.
func RWTopologyFromJSONBytes(b []byte) (*RWTopology, error) {
	rt := &jsontopo.Topology{}
	if err := json.Unmarshal(b, rt); err != nil {
		return nil, err
	}
	ct, err := RWTopologyFromJSONTopology(rt)
	if err != nil {
		return nil, serrors.WrapStr("unable to convert raw topology to topology", err)
	}
	return ct, nil
}

// RWTopologyFromJSONFile extracts the topology from a file containing the JSON representation
// of the topology.
func RWTopologyFromJSONFile(path string) (*RWTopology, error) {
	b, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return RWTopologyFromJSONBytes(b)
}

func (t *RWTopology) populateMeta(raw *jsontopo.Topology) error {
	// These fields can be simply copied
	var err error
	t.Timestamp = time.Unix(raw.Timestamp, 0)

	if t.IA, err = addr.IAFromString(raw.IA); err != nil {
		return err
	}
	if t.IA.IsWildcard() {
		return serrors.New("ISD-AS contains wildcard", "isd_as", t.IA)
	}
	t.MTU = raw.MTU
	t.Attributes = raw.Attributes
	return nil
}

func (t *RWTopology) populateBR(raw *jsontopo.Topology) error {
	for name, rawBr := range raw.BorderRouters {
		if rawBr.CtrlAddr == "" {
			return serrors.New("Missing Control Address", "br", name)
		}
		if rawBr.InternalAddr == "" {
			return serrors.New("Missing Internal Address", "br", name)
		}
		ctrlAddr, err := rawAddrToTopoAddr(rawBr.CtrlAddr)
		if err != nil {
			return serrors.WrapStr("unable to extract SCION control-plane address", err)
		}
		intAddr, err := rawAddrToUDPAddr(rawBr.InternalAddr)
		if err != nil {
			return serrors.WrapStr("unable to extract underlay internal data-plane address", err)
		}
		brInfo := BRInfo{
			Name:         name,
			CtrlAddrs:    ctrlAddr,
			InternalAddr: intAddr,
			IFs:          make(map[common.IFIDType]*IFInfo),
		}
		for ifid, rawIntf := range rawBr.Interfaces {
			var err error
			// Check that ifid is unique
			if _, ok := t.IFInfoMap[ifid]; ok {
				return serrors.New("IFID already exists", "ID", ifid)
			}
			brInfo.IFIDs = append(brInfo.IFIDs, ifid)
			ifinfo := IFInfo{
				ID:           ifid,
				BRName:       name,
				InternalAddr: intAddr,
				CtrlAddrs:    ctrlAddr,
				Bandwidth:    rawIntf.Bandwidth,
				MTU:          rawIntf.MTU,
			}
			if ifinfo.IA, err = addr.IAFromString(rawIntf.IA); err != nil {
				return err
			}
			ifinfo.LinkType = LinkTypeFromString(rawIntf.LinkTo)
			isCore := false
			for _, attr := range t.Attributes {
				if attr == jsontopo.AttrCore {
					isCore = true
					break
				}
			}
			if err = ifinfo.CheckLinks(isCore, name); err != nil {
				return err
			}
			if bfd := rawIntf.BFD; bfd != nil {
				ifinfo.BFD = BFD{
					Disable:               bfd.Disable,
					DetectMult:            bfd.DetectMult,
					DesiredMinTxInterval:  bfd.DesiredMinTxInterval.Duration,
					RequiredMinRxInterval: bfd.RequiredMinRxInterval.Duration,
				}
			}

			// These fields are only necessary for the border router.
			// Parsing should not fail if they are missing.
			if rawIntf.Underlay.Bind == "" && rawIntf.Underlay.Remote == "" {
				brInfo.IFs[ifid] = &ifinfo
				t.IFInfoMap[ifid] = ifinfo
				continue
			}
			if ifinfo.Local, err = rawBRIntfTopoBRAddr(rawIntf); err != nil {
				return serrors.WrapStr("unable to extract "+
					"underlay external data-plane local address", err)
			}
			if ifinfo.Remote, err = rawAddrToUDPAddr(rawIntf.Underlay.Remote); err != nil {
				return serrors.WrapStr("unable to extract "+
					"underlay external data-plane remote address", err)
			}
			ifinfo.Underlay = underlay.UDPIPv6
			if ifinfo.Local.IP.To4() != nil && ifinfo.Remote.IP.To4() != nil {
				ifinfo.Underlay = underlay.UDPIPv4
			}
			brInfo.IFs[ifid] = &ifinfo
			t.IFInfoMap[ifid] = ifinfo
		}
		sort.Slice(brInfo.IFIDs, func(i, j int) bool {
			return brInfo.IFIDs[i] < brInfo.IFIDs[j]
		})
		t.BR[name] = brInfo
		t.BRNames = append(t.BRNames, name)
	}
	sort.Strings(t.BRNames)
	return nil
}

func (t *RWTopology) populateServices(raw *jsontopo.Topology) error {
	var err error
	t.CS, err = svcMapFromRaw(raw.ControlService)
	if err != nil {
		return serrors.WrapStr("unable to extract CS address", err)
	}
	t.SIG, err = gatewayMapFromRaw(raw.SIG)
	if err != nil {
		return serrors.WrapStr("unable to extract SIG address", err)
	}
	t.DS, err = svcMapFromRaw(raw.DiscoveryService)
	if err != nil {
		return serrors.WrapStr("unable to extract DS address", err)
	}
	t.HiddenSegmentLookup, err = svcMapFromRaw(raw.HiddenSegmentLookup)
	if err != nil {
		return serrors.WrapStr("unable to extract hidden segment lookup address", err)
	}
	t.HiddenSegmentRegistration, err = svcMapFromRaw(raw.HiddenSegmentReg)
	if err != nil {
		return serrors.WrapStr("unable to extract hidden segment registration address", err)
	}
	return nil
}

// Active returns whether the topology is active at the point in time specified by the argument.
// A topology is active if now is after the timestamp.
func (t *RWTopology) Active(now time.Time) bool {
	return !now.Before(t.Timestamp)
}

// GetTopoAddr returns the address information for the process of the requested type with the
// requested ID.
func (t *RWTopology) GetTopoAddr(id string, svc ServiceType) (*TopoAddr, error) {
	svcInfo, err := t.getSvcInfo(svc)
	if err != nil {
		return nil, err
	}
	topoAddr := svcInfo.idTopoAddrMap.GetByID(id)
	if topoAddr == nil {
		return nil, serrors.New("Element not found", "id", id)
	}
	return topoAddr, nil
}

// getAllTopoAddrs returns the address information of all processes of the requested type.
func (t *RWTopology) getAllTopoAddrs(svc ServiceType) ([]TopoAddr, error) {
	svcInfo, err := t.getSvcInfo(svc)
	if err != nil {
		return nil, err
	}
	topoAddrs := svcInfo.getAllTopoAddrs()
	if topoAddrs == nil {
		return nil, ErrAddressNotFound
	}
	return topoAddrs, nil
}

func (t *RWTopology) getSvcInfo(svc ServiceType) (*svcInfo, error) {
	switch svc {
	case Unknown:
		return nil, serrors.New("service type unknown")
	case Discovery:
		return &svcInfo{idTopoAddrMap: t.DS}, nil
	case Control:
		return &svcInfo{idTopoAddrMap: t.CS}, nil
	case HiddenSegmentLookup:
		return &svcInfo{idTopoAddrMap: t.HiddenSegmentLookup}, nil
	case HiddenSegmentRegistration:
		return &svcInfo{idTopoAddrMap: t.HiddenSegmentRegistration}, nil
	case Gateway:
		m := make(IDAddrMap)
		for k, v := range t.SIG {
			m[k] = *v.CtrlAddr
		}
		return &svcInfo{idTopoAddrMap: m}, nil
	default:
		return nil, serrors.New("unsupported service type", "type", svc)
	}
}

// Copy creates a deep copy of the object.
func (t *RWTopology) Copy() *RWTopology {
	if t == nil {
		return nil
	}
	return &RWTopology{
		Timestamp:  t.Timestamp,
		IA:         t.IA,
		MTU:        t.MTU,
		Attributes: append(t.Attributes[:0:0], t.Attributes...),

		BR:        copyBRMap(t.BR),
		BRNames:   append(t.BRNames[:0:0], t.BRNames...),
		IFInfoMap: t.IFInfoMap.copy(),

		CS:                        t.CS.copy(),
		DS:                        t.DS.copy(),
		SIG:                       copySIGMap(t.SIG),
		HiddenSegmentLookup:       t.HiddenSegmentLookup.copy(),
		HiddenSegmentRegistration: t.HiddenSegmentRegistration.copy(),
	}
}

func copySIGMap(m map[string]GatewayInfo) map[string]GatewayInfo {
	if m == nil {
		return nil
	}
	ret := make(map[string]GatewayInfo)
	for k, v := range m {
		e := GatewayInfo{
			CtrlAddr: v.CtrlAddr.copy(),
			DataAddr: copyUDPAddr(v.DataAddr),
		}
		ret[k] = e
	}
	return ret
}

func copyBRMap(m map[string]BRInfo) map[string]BRInfo {
	if m == nil {
		return nil
	}
	newM := make(map[string]BRInfo)
	for k, v := range m {
		newM[k] = *v.copy()
	}
	return newM
}

func (i *BRInfo) copy() *BRInfo {
	if i == nil {
		return nil
	}
	return &BRInfo{
		Name:         i.Name,
		CtrlAddrs:    i.CtrlAddrs.copy(),
		InternalAddr: copyUDPAddr(i.InternalAddr),
		IFIDs:        append(i.IFIDs[:0:0], i.IFIDs...),
		IFs:          copyIFsMap(i.IFs),
	}
}

func copyIFsMap(m map[common.IFIDType]*IFInfo) map[common.IFIDType]*IFInfo {
	if m == nil {
		return nil
	}
	newM := make(map[common.IFIDType]*IFInfo)
	for k, v := range m {
		newM[k] = v.copy()
	}
	return newM
}

func (m IfInfoMap) copy() IfInfoMap {
	if m == nil {
		return nil
	}
	newM := make(IfInfoMap)
	for k, v := range m {
		newM[k] = *v.copy()
	}
	return newM
}

// svcInfo contains topology information for a single SCION service
type svcInfo struct {
	idTopoAddrMap IDAddrMap
}

func (svc *svcInfo) getAllTopoAddrs() []TopoAddr {
	var topoAddrs []TopoAddr
	for _, topoAddr := range svc.idTopoAddrMap {
		topoAddrs = append(topoAddrs, topoAddr)
	}
	return topoAddrs
}

func svcMapFromRaw(ras map[string]*jsontopo.ServerInfo) (IDAddrMap, error) {
	svcMap := make(IDAddrMap)
	for name, svc := range ras {
		svcTopoAddr, err := rawAddrToTopoAddr(svc.Addr)
		if err != nil {
			return nil, serrors.WrapStr("could not parse address", err,
				"address", svc.Addr, "process_name", name)
		}
		svcMap[name] = *svcTopoAddr
	}
	return svcMap, nil
}

func gatewayMapFromRaw(ras map[string]*jsontopo.GatewayInfo) (map[string]GatewayInfo, error) {
	ret := make(map[string]GatewayInfo)
	for name, svc := range ras {
		c, err := rawAddrToTopoAddr(svc.CtrlAddr)
		if err != nil {
			return nil, serrors.WrapStr("could not parse control address", err,
				"address", svc.CtrlAddr, "process_name", name)
		}
		d, err := rawAddrToUDPAddr(svc.DataAddr)
		if err != nil {
			return nil, serrors.WrapStr("could not parse data address", err,
				"address", svc.DataAddr, "process_name", name)
		}

		ret[name] = GatewayInfo{
			CtrlAddr:        c,
			DataAddr:        d,
			AllowInterfaces: svc.Interfaces,
		}
	}
	return ret, nil
}

// GetByID returns the TopoAddr for the given ID, or nil if there is none.
func (m IDAddrMap) GetByID(id string) *TopoAddr {
	if _, ok := m[id]; ok {
		cp := m[id]
		return &cp
	}
	return nil
}

func (m IDAddrMap) copy() IDAddrMap {
	if m == nil {
		return nil
	}
	newM := make(IDAddrMap)
	for k, v := range m {
		// This has the potential of making a _lot_ of shallow copies, but we can't really avoid it
		// due to the value type in the map.
		newM[k] = *v.copy()
	}
	return newM
}

// CheckLinks checks whether the link types are compatible with whether the AS is core or not.
func (i IFInfo) CheckLinks(isCore bool, brName string) error {
	if isCore {
		switch i.LinkType {
		case Core, Child:
		default:
			return serrors.New("Illegal link type for core AS",
				"type", i.LinkType, "br", brName)
		}
	} else {
		switch i.LinkType {
		case Parent, Child, Peer:
		default:
			return serrors.New("Illegal link type for non-core AS",
				"type", i.LinkType, "br", brName)
		}
	}
	return nil
}

func (i IFInfo) String() string {
	return fmt.Sprintf("IFinfo: Name[%s] IntAddr[%+v] CtrlAddr[%+v] Underlay:%s Local:%+v "+
		"Remote:%+v Bw:%d IA:%s Type:%v MTU:%d", i.BRName, i.InternalAddr, i.CtrlAddrs, i.Underlay,
		i.Local, i.Remote, i.Bandwidth, i.IA, i.LinkType, i.MTU)
}

func (i *IFInfo) copy() *IFInfo {
	if i == nil {
		return nil
	}
	return &IFInfo{
		ID:           i.ID,
		BRName:       i.BRName,
		CtrlAddrs:    i.CtrlAddrs.copy(),
		Underlay:     i.Underlay,
		InternalAddr: copyUDPAddr(i.InternalAddr),
		Local:        copyUDPAddr(i.Local),
		Remote:       copyUDPAddr(i.Remote),
		RemoteIFID:   i.RemoteIFID,
		Bandwidth:    i.Bandwidth,
		IA:           i.IA,
		LinkType:     i.LinkType,
		MTU:          i.MTU,
	}
}

// UnderlayAddr returns the underlay address interpreted as a net.UDPAddr.
//
// FIXME(scrye): This should be removed; applications should not need to look into the underlay
// concrete type.
func (a *TopoAddr) UnderlayAddr() *net.UDPAddr {
	return a.UnderlayAddress
}

func (a *TopoAddr) String() string {
	return fmt.Sprintf("TopoAddr{SCION: %v, Underlay: %v}", a.SCIONAddress, a.UnderlayAddress)
}

func (a *TopoAddr) copy() *TopoAddr {
	// TODO(scrye): Investigate how this can be removed.
	if a == nil {
		return nil
	}
	return &TopoAddr{
		SCIONAddress:    copyUDPAddr(a.SCIONAddress),
		UnderlayAddress: toUDPAddr(a.UnderlayAddress),
	}
}

func toUDPAddr(a net.Addr) *net.UDPAddr {
	if a == nil {
		return nil
	}
	udpAddr, ok := a.(*net.UDPAddr)
	if !ok {
		return nil
	}
	return copyUDPAddr(udpAddr)
}

// ServiceNames is a slice of process names (e.g., "bs-1", "bs-2").
type ServiceNames []string

// GetRandom returns a random entry, or an error if the slice is empty.
func (s ServiceNames) GetRandom() (string, error) {
	numServers := len(s)
	if numServers == 0 {
		return "", serrors.New("No names present")
	}
	return s[rand.Intn(numServers)], nil
}

func copyUDPAddr(a *net.UDPAddr) *net.UDPAddr {
	if a == nil {
		return nil
	}
	return &net.UDPAddr{
		IP:   append(a.IP[:0:0], a.IP...),
		Port: a.Port,
		Zone: a.Zone,
	}
}
